src/main/java/jp/co/soramitsu/sora/common/ArrayTree.java

```diff
@@ -0,0 +1,96 @@
+package jp.co.soramitsu.sora.common;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.Getter;
+
+public class ArrayTree<E> {
+
+  @Getter
+  private List<E> tree;
+
+  public ArrayTree(List<E> tree) {
+    this.tree = tree;
+  }
+
+  public boolean hasElement(int i) {
+    return i < tree.size() && get(i) != null;
+  }
+
+  public List<E> getLeafs() {
+    int leafsTotal = (tree.size() + 1) / 2;
+    int leftmostLeafIndex = leafsTotal - 1;
+    return tree.stream().skip(leftmostLeafIndex).collect(Collectors.toList());
```

Comment:
as this class is immutable we can have leaves stored in specific field, and just return them in this method, allowing us to lower processing time

Reply:
This class is not immutable